### ISSUE TITLE
JSONPath.toPathArray returns different values for the same input

### ISSUE THREAD

**Issue by andalal** [JSONPath-Plus/JSONPath#102]
I am using the toPathArray method in the following way

`const jsonpathTokens: string[] = JSONPath.toPathArray(propertyRef.propertyPath);`

When I pass in the string `"$.subsensor.temp"` for the first time, it will return `["$", "subsensor", "temp"]` as expected. However when the toPathArray is called again with the same input, it returns `["subsensor", "temp"]`. This happens for other inputs of a similar type.

When I stepped through the debugger it seems like there is a JSONPath.cache object which is checked before the regex is run. I checked that the value saved to the cache `exprList` is correct. However on the immediate next call the cache object was already modified to exclude the "$". 

I think this is a bug because the toPathArray method should return the same values for the same input. 

I am using v1.0.0 and I'm using the library within typescript.

**Comment by brettz9** [comment]
Are you able to replicate in plain JavaScript?

Running the following in Node (if that is your environment), shows no differences:

```js
const {JSONPath} = require('./');

const a = JSONPath.toPathArray('$.subsensor.temp');

console.log('jsonpath', JSON.stringify(a), JSONPath.cache);

const b = JSONPath.toPathArray('$.subsensor.temp');

console.log('jsonpath', JSON.stringify(b), JSONPath.cache);
```

And this in the browser:

```html
<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <title>Test</title>
  </head>
  <body>
    <script type="module">
      import {JSONPath} from './dist/index-es.js';

      const a = JSONPath.toPathArray('$.subsensor.temp');

      console.log('jsonpath', JSON.stringify(a), JSONPath.cache);

      const b = JSONPath.toPathArray('$.subsensor.temp');

      console.log('jsonpath', JSON.stringify(b), JSONPath.cache);
    </script>
  </body>
</html>
```

**Comment by bjornharrtell** [comment]
I'm seeing the same issue, setting JSONPath.cache = {} before calling JSONPath.toPathArray will work around it.

**Comment by brettz9** [comment]
Can you prepare a test case, @bjornharrtell ?

**Comment by bjornharrtell** [comment]
I will try. :)

**Comment by tejodorus** [comment]
I have the issue and likely the cause. At the end of toPathArray, the plain cached object is returned:
return cache[expr]
but it should be a clone (return cache[expr].concat()), like before in that function "if (cache[expr]) { return cache[expr].concat();"

This toPathArray is used within "evaluate", and there the result is shifted, which removes the first element.

So, the proposed solution is to return cache[expr].concat()

**Comment by sdolski** [comment]
I am facing the same problem and also tracked it down to the cache.
I can confirm that the analysis, as well as the proposed solution of @tejodorus is correct.

Any chance of geting the linked PR merged?

**Comment by brettz9** [comment]
Released in v5.0.5. Thanks for the report!
